File: GAS/src/fi/quanfoxes/Parser/patterns/OperatorPattern.java
package fi.quanfoxes.Parser.patterns;

import fi.quanfoxes.Lexer.OperatorToken;
import fi.quanfoxes.Lexer.Token;
import fi.quanfoxes.Lexer.TokenType;
import fi.quanfoxes.Parser.Node;
import fi.quanfoxes.Parser.Pattern;
import fi.quanfoxes.Parser.nodes.*;

import java.util.List;

public class OperatorPattern extends Pattern {
    private static final int LEFT = 0;
    private static final int OPERATOR = 1;
    private static final int RIGHT = 2;

    public OperatorPattern() {
        super(TokenType.IDENTIFIER | TokenType.NUMBER | TokenType.PROCESSED, TokenType.OPERATOR,
                    TokenType.IDENTIFIER | TokenType.NUMBER | TokenType.PROCESSED);
    }

    @Override
    public int priority(List<Token> tokens) {
        final OperatorToken operator = (OperatorToken)tokens.get(OPERATOR);
        return operator.getOperator().getPriority();
    }

    @Override
    public boolean passes(List<Token> tokens) {
        return true;
    }

    @Override
    public Node build(Node parent, List<Token> tokens) throws Exception {
        Token left = tokens.get(LEFT);
        OperatorToken operator = (OperatorToken)tokens.get(OPERATOR);
        Token right = tokens.get(RIGHT);

        return new OperatorNode((ContextNode) parent, left, operator.getOperator(), right);
    }
}

File: Vivid/Parser/Patterns/OperatorPattern.cs
using System.Collections.Generic;

public class OperatorPattern : Pattern
{
	private const int LEFT = 0;
	private const int OPERATOR = 2;
	private const int RIGHT = 4;

	// Pattern: ... [\n] $operator [\n] ...
	public OperatorPattern() : base
	(
		TokenType.OBJECT,
		TokenType.END | TokenType.OPTIONAL,
		TokenType.OPERATOR,
		TokenType.END | TokenType.OPTIONAL,
		TokenType.OBJECT
	)
	{ Priority = Parser.PRIORITY_ALL; }

	public override bool Passes(Context context, ParserState state, List<Token> tokens, int priority)
	{
		return tokens[OPERATOR].To<OperatorToken>().Operator.Priority == priority;
	}

	public override Node Build(Context context, ParserState state, List<Token> tokens)
	{
		return new OperatorNode(tokens[OPERATOR].To<OperatorToken>().Operator, tokens[OPERATOR].Position).SetOperands(
			Singleton.Parse(context, tokens[LEFT]),
			Singleton.Parse(context, tokens[RIGHT])
		);
	}
}